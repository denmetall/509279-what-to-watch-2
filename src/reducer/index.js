import {combineReducers} from 'redux';

import {reducer as genre} from './genre/genre';
import {reducer as films} from './films/films';
import {reducer as authorization} from './authorization/authorization';
import {reducer as reviews} from './reviews/reviews';
import {reducer as favorites} from './favorites/favorites';

export default combineReducers(
    {
      genre,
      films,
      authorization,
      reviews,
      favorites
    }
);
