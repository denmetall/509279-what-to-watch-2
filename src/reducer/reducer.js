import {combineReducers} from 'redux';

import {reducer as genre} from './genre/genre';
import {reducer as films} from './films/films';
import {reducer as authorization} from './authorization/authorization';
import {reducer as reviews} from './reviews/reviews';
import {reducer as favorites} from './favorites/favorites';
import {reducer as networkFailed} from './network-failed/network-failed';
import {reducer as loadersReducer} from "react-redux-hoc-loader";

export default combineReducers(
    {
      genre,
      films,
      authorization,
      reviews,
      favorites,
      networkFailed,
      loaders: loadersReducer,
    }
);
