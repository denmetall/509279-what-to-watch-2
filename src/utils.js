import {shape, string, number, arrayOf, bool} from 'prop-types';

const Tabs = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`
};

const CommentLength = {
  MIN: 50,
  MAX: 400
};

const MAX_RATING = 5;

const BASE_URL_SERVER = `https://htmlacademy-react-2.appspot.com`;

const API = {
  BASE_URL: `https://htmlacademy-react-2.appspot.com/wtw`,
  TIMEOUT: 5000,
  WITH_CREDENTIALS: true
};

const MoviesCount = {
  DEFAULT: 8,
  STEP: 20
};

const DEFAULT_FILTER = `All genres`;

const MovieType = shape({
  id: number,
  name: string,
  description: string,
  genre: string,
  rating: number,
  scoresCount: number,
  director: string,
  starring: arrayOf(string),
  runTime: number,
  released: number,
  isFavorite: bool,
  posterImage: string,
  previewImage: string,
  backgroundImage: string,
  backgroundColor: string,
  videoLink: string,
  previewVideoLink: string
});

const UserType = shape({
  id: number,
  email: string,
  name: string,
  avatarUrl: string,
});

const ReviewType = shape({
  id: number,
  user: shape({
    id: number,
    name: string
  }),
  rating: number,
  comment: string,
  date: string
});

const Ratings = {
  ONE: 1,
  TWO: 2,
  THREE: 3,
  FOUR: 4,
  FIVE: 5
};

export {
  Tabs,
  DEFAULT_FILTER,
  API,
  BASE_URL_SERVER,
  MovieType,
  UserType,
  ReviewType,
  CommentLength,
  MAX_RATING,
  MoviesCount,
  Ratings
};
