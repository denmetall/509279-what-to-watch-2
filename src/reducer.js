import filmsMock from './mocks/films';
import {DEFAULT_FILTER} from "./utils";

const initialState = {
  genre: DEFAULT_FILTER,
  films: filmsMock,
  filmsSort: filmsMock,
  promoFilm: filmsMock[0]
};

const ActionType = {
  SET_FILTER_GENRE: `SET_FILTER_GENRE`,
  GET_MOVIES_LIST: `GET_MOVIES_LIST`
};

const getFilterMovies = (moviesGenre) => {
  switch (moviesGenre) {
    case DEFAULT_FILTER:
      return filmsMock;
  }

  return filmsMock.slice().filter((movie) => movie.genre === moviesGenre);
};

const ActionCreator = {
  setGenreFilter: (genre) => ({
    type: ActionType.SET_FILTER_GENRE,
    payload: genre
  }),

  getMoviesList: (genre) => ({
    type: ActionType.GET_MOVIES_LIST,
    payload: getFilterMovies(genre)
  })
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_FILTER_GENRE:
      return Object.assign({}, state, {
        genre: action.payload
      });
    case ActionType.GET_MOVIES_LIST:
      return Object.assign({}, state, {
        filmsSort: action.payload
      });
  }

  return state;
};

export {
  reducer,
  ActionCreator,
  ActionType
};
