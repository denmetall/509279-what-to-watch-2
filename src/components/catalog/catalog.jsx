import React, {PureComponent} from 'react';
import {connect} from 'react-redux';
import GenresList from "../genres-list/genres-list.jsx";
import MovieCardsList from "../movie-cards-list/movie-cards-list.jsx";
import ShowMore from "../show-more/show-more.jsx";
import PropTypes from "prop-types";
import {ActionCreator} from "../../reducer";
import {MOVIES_COUNT_DEFAULT, MOVIES_COUNT_STEP, DEFAULT_FILTER} from "../../utils";

class Catalog extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      moviesCounter: MOVIES_COUNT_DEFAULT
    };

    this._onShowMoreClick = this._onShowMoreClick.bind(this);
  }

  render() {
    const {films, genre, onChangeFilter} = this.props;

    return <section className="catalog">
      <h2 className="catalog__title visually-hidden">Catalog</h2>

      <GenresList activeFilter={genre} onChangeFilter={onChangeFilter}/>

      <MovieCardsList films={films.slice(0, this.state.moviesCounter)}/>

      <div className="catalog__more">
        <ShowMore onClick={this._onShowMoreClick}/>
      </div>
    </section>;
  }

  _onShowMoreClick() {
    this.setState((prevState) => ({
      moviesCounter: prevState.moviesCounter + MOVIES_COUNT_STEP
    }));
  }
}

Catalog.defaultProps = {
  films: [],
  genre: ``
};

Catalog.propTypes = {
  films: PropTypes.arrayOf(PropTypes.object).isRequired,
  genre: PropTypes.string.isRequired,
  onChangeFilter: PropTypes.func.isRequired
};

const mapStateToProps = (state) => {
  return {
    genre: state.genre,
    films: state.films.filter((movie) => {
      if (state.genre === DEFAULT_FILTER) {
        return true;
      } else {
        return movie.genre === state.genre;
      }
    })
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onChangeFilter: (genre) => {
      dispatch(ActionCreator.setGenreFilter(genre));
    }
  };
};

export {Catalog};

export default connect(mapStateToProps, mapDispatchToProps)(Catalog);
