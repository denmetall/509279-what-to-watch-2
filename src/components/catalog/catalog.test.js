import React from 'react';
import {Catalog} from './catalog.jsx';
import renderer from 'react-test-renderer';
import filmsMock from '../../mocks/films';
import {HashRouter as Router} from 'react-router-dom';
import createNodeMock from '../../mocks/create-node-mock';
import {DEFAULT_FILTER} from "../../utils";
import {createStore} from 'redux';
import {reducer} from '../../reducer';
import {Provider} from 'react-redux';

it(`Components BtnPlay renders correctly`, () => {
  const options = {createNodeMock};
  const store = createStore(reducer);
  const tree = renderer
    .create(<Provider store={store}>
      <Router>
        <Catalog
          films={filmsMock}
          filmsSort={filmsMock}
          genre={DEFAULT_FILTER}
          onChangeFilter={jest.fn()}
        />
      </Router>
    </Provider>, options)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
