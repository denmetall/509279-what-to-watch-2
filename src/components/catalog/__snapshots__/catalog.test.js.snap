// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components BtnPlay renders correctly 1`] = `
<section
  className="catalog"
>
  <h2
    className="catalog__title visually-hidden"
  >
    Catalog
  </h2>
  <ul
    className="catalog__genres-list"
  >
    <li
      className="catalog__genres-item catalog__genres-item--active"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        All genres
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Triller
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Horror
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Comedy
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Militant
      </a>
    </li>
  </ul>
  <div
    className="catalog__movies-list"
  >
    <article
      className="small-movie-card catalog__movies-card"
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <video
          height="175"
          muted={true}
          poster="img/johnny-english.jpg"
          width="280"
        >
          <source
            src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
          />
        </video>
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="#/film/1"
          onClick={[Function]}
        >
          Drive
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <video
          height="175"
          muted={true}
          poster="img/macbeth.jpg"
          width="280"
        >
          <source
            src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
          />
        </video>
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="#/film/2"
          onClick={[Function]}
        >
          Astral
        </a>
      </h3>
    </article>
  </div>
  <div
    className="catalog__more"
  >
    <button
      className="catalog__button"
      onClick={[Function]}
      type="button"
    >
      Show more
    </button>
  </div>
</section>
`;
